rclrs/src/node/client.rs

```diff
@@ -0,0 +1,257 @@
+// they are running in. Therefore, this type can be safely sent to another thread.
+unsafe impl Send for rcl_client_t {}
+
+/// Internal struct used by clients.
+pub struct ClientHandle {
+    rcl_client_mtx: Mutex<rcl_client_t>,
+    rcl_node_mtx: Arc<Mutex<rcl_node_t>>,
+}
+
+impl ClientHandle {
+    pub(crate) fn lock(&self) -> MutexGuard<rcl_client_t> {
+        self.rcl_client_mtx.lock()
+    }
+}
+
+impl Drop for ClientHandle {
+    fn drop(&mut self) {
+        let handle = self.rcl_client_mtx.get_mut();
+        let rcl_node_mtx = &mut *self.rcl_node_mtx.lock();
+        unsafe {
+            rcl_client_fini(handle as *mut _, rcl_node_mtx as *mut _);
+        }
+    }
+}
+
+impl From<Canceled> for RclrsError {
+    fn from(_: Canceled) -> Self {
+        RclrsError::RclError {
+            code: RclReturnCode::Error,
+            msg: None,
+        }
+    }
+}
+
+/// Trait to be implemented by concrete Client structs
+/// See [`Client<T>`] for an example
+pub trait ClientBase: Send + Sync {
+    /// Internal function to get a reference to the `rcl` handle.
+    fn handle(&self) -> &ClientHandle;
+    /// Tries to take a new response and run the callback or future with it.
+    fn execute(&self) -> Result<(), RclrsError>;
+}
+
+type RequestValue<Response> = Box<dyn FnOnce(&Response) + 'static + Send>;
+
+/// Main class responsible for publishing data to ROS topics
+pub struct Client<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    pub(crate) handle: Arc<ClientHandle>,
+    requests: Mutex<HashMap<i64, RequestValue<T::Response>>>,
+    futures: Arc<Mutex<HashMap<i64, oneshot::Sender<T::Response>>>>,
+    sequence_number: AtomicI64,
+}
+
+impl<T> Client<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    /// Creates a new client.
+    pub fn new(node: &Node, topic: &str) -> Result<Self, RclrsError>
+    where
+        T: rosidl_runtime_rs::Service,
+    {
+        let mut rcl_client = unsafe { rcl_get_zero_initialized_client() };
+        let type_support = <T as rosidl_runtime_rs::Service>::get_type_support()
+            as *const rosidl_service_type_support_t;
+        let topic_c_string = CString::new(topic).unwrap();
+        let rcl_node = &mut *node.rcl_node_mtx.lock();
+
+        unsafe {
+            let client_options = rcl_client_get_default_options();
+
+            rcl_client_init(
+                &mut rcl_client,
+                rcl_node,
+                type_support,
+                topic_c_string.as_ptr(),
+                &client_options,
+            )
+            .ok()?;
+        }
+
+        let handle = Arc::new(ClientHandle {
+            rcl_client_mtx: Mutex::new(rcl_client),
+            rcl_node_mtx: node.rcl_node_mtx.clone(),
+        });
+
+        Ok(Self {
+            handle,
+            requests: Mutex::new(HashMap::new()),
+            futures: Arc::new(Mutex::new(
+                HashMap::<i64, oneshot::Sender<T::Response>>::new(),
```

Comment:
You could just make this `HashMap::new()` if you want, I believe. But this way is also fine.